### PR TITLE
Implement extensions functions to make use of AL_EXT_float32 enums

### PR DESCRIPTION
**Pull request by tilkinsc**
# Summary of the PR
Currently you have to explicitly cast FloatBufferFormat to BufferFormat to use the AL::BufferFormat function. alBufferFormat takes in two new enums provided by AL_EXT_float32  to provide the ability to load floating point data into openal buffer objects.

# Related issues, Discord discussions, or proposals
Issue Ref: #1149
[Spec Ref](https://github.com/openalext/openalext/blob/master/AL_EXT_float32.txt)

# Further Comments
There may be a possibility that the Capture feature of OpenAL supports floating point audio. I have no mind to include this because it adds more research that isn't easily presented in google. Silk.NET doesn't compile too well either. So, to keep it simple I won't add anything to do with anything other than what the spec AL_EXT_float32 says.

### CODE REVIEW

**Review by Beyley — APPROVED** [review]

**Comment by Beyley** [comment]
Want to get another maintainer's opinion before merging, as i dont think theres precedent for this kind of overload on extensions, and it only really applies to the manually generated bindings (like OpenAL)

**Review by HurricanKai — APPROVED** [review]

**Comment by tilkinsc** [comment]
I want to add a forethought, it appears that OpenAL-soft doesn't provide functionality for AL_EXT_vorbis and AL_EXT_mp3.

https://github.com/kcat/openal-soft/blob/master/al/buffer.cpp
https://github.com/kcat/openal-soft/blob/master/include/AL/alext.h

So by all means only AL_EXT_float32, AL_EXT_IMA4, AL_EXT_double, AL_EXT_MULAW, AL_EXT_ALAW would ideally need toyed with in the future for alBufferData

THAT IS,

if there was some friggin documentation on what to pass INTO alBufferData. Do I give it the whole mp3 file?! We will never know.

**Comment by Beyley** [comment]
> I want to add a forethought, it appears that OpenAL-soft doesn't provide functionality for AL_EXT_vorbis and AL_EXT_mp3.
> 
> https://github.com/kcat/openal-soft/blob/master/al/buffer.cpp https://github.com/kcat/openal-soft/blob/master/include/AL/alext.h
> 
> So by all means only AL_EXT_float32, AL_EXT_IMA4, AL_EXT_double, AL_EXT_MULAW, AL_EXT_ALAW would ideally need toyed with in the future for alBufferData
> 
> THAT IS,
> 
> if there was some friggin documentation on what to pass INTO alBufferData. Do I give it the whole [insert file format here] file?! We will never know.

@ohtrobinson you may be able to provide explination here on what `alBufferData` wants, i believe i know, but you may be able to explain better than i can, given you have more familiarity than i do

**Comment by ohtrobinson** [comment]
I'm not too sure myself, I've not experimented with it before. I can't find where I heard it, but I believe that (taking vorbis as an example) it specifically wants vorbis *sound* data, and you need to parse the headers yourself. All it can do is process the given data into a PCM stream, and doesn't do any actual loading.

There really isn't that much about these extensions anywhere online, and as such I would recommend avoiding them, instead opting to use StbVorbis and an MP3 loader. At least then you can guarantee that it will work, as well as working on machines that don't support the extension.

I'm surprised OpenAL soft doesn't appear to support vorbis, but I sorta understand, it's a royal pain to work with, especially in the way OpenAL works.

**Comment by tilkinsc** [comment]
SoLoud is implemented with loading as well as playing audio. I would trial it to be a useful addition to sit alongside OpenAL, being that it has the zlib (basically MIT) license. They sure as hell did a good job keeping it free software in that regard too.

SoLoud: zlib/libpng
OGG: public domain
MP3: Unlicense
Wav: Unlicense
Speech Synth: Public domain (including author changes)
Vizsn (speech synth): Public domain
FFT: public domain
Sfxr: zlib/libpng
RPGMaker Wrapper Generator: BSDL (never claimed to be part of Ruby) but actually public domain
TED/SID soundchip emulation: public domain
Vic soundchip emulation: zlib/libpng
Openmpt (completely optional): BSD-3 https://opensource.org/licenses/BSD-3-Clause
FreeVerb: public domain

http://solhsa.com/soloud/legal.html

Would this be an interesting proposal?

**Comment by Perksey** [comment]
We generally try to stick to well-known, industry-standard APIs specified by authoritative bodies for silk-provided bindings, with the exception of the HLU which binds to whatever library necessary to deliver its managed abstractions.

The audio space is very fragmented, heck to be honest it’s questionable why we even have OpenAL bindings given it’s not nearly as recognised or operated in the same way as their graphics counterparts. As such, I do not believe Silk.NET would have SoLoud bindings given this seems like we’re “picking” a preferred technology, and making our selection of bindings more opinionated.

However, the SilkCommunity has no such rules. If someone made SoLoud bindings/wrappers/utilities using our tech (be that our generator tech and/or Silk.NET.OpenAL itself to “tie” those libraries together) I’m sure the SilkCommunity would love to see it.

https://GitHub.com/SilkCommunity/SilkCommunity

**Comment by Beyley** [comment]
> We generally try to stick to well-known, industry-standard APIs specified by authoritative bodies for silk-provided bindings, with the exception of the HLU which binds to whatever library necessary to deliver its managed abstractions.
> 
> The audio space is very fragmented, heck to be honest it’s questionable why we even have OpenAL bindings given it’s not nearly as recognised or operated in the same way as their graphics counterparts. As such, I do not believe Silk.NET would have SoLoud bindings given this seems like we’re “picking” a preferred technology, and making our selection of bindings more opinionated.
> 
> However, the SilkCommunity has no such rules. If someone made SoLoud bindings/wrappers/utilities using our tech (be that our generator tech and/or Silk.NET.OpenAL itself to “tie” those libraries together) I’m sure the SilkCommunity would love to see it.
> 
> https://GitHub.com/SilkCommunity/SilkCommunity

To add to this, the generator is 95% of the way there to be able to generate SoLoud bindings (i believe there were a few random problems, but it mostly worked), so if someone wants to make them, @ me on discord, and ill help out best i can

**Pull request merged**
